Make serializable classes less fragile to small build and implementation changes

ldziedziul opened the pull request:
Related to https://github.com/hazelcast/hazelcast/issues/19320

- Added `serialVersionUID` fields to `Transform` implementation in Jet and in serializable classes in `hazelcast-jet-kinesis` extension
- Added `hazelcast-archunit-rules` module with a shareable ArchUnit rule for Serializable classes
- A [test](https://github.com/hazelcast/hazelcast/pull/19534/files#diff-8723ab58d30463912f118b197f6d873a11321470e207b8e404d14844b9323f18R27) against the ArchUnit rule was added only for `hazelcast-jet-kinesis` since other modules need to be fixed first.

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Add `Add to Release Notes` label if changes should be mentioned in release notes or `Not Release Notes content` if changes are not relevant for release notes
- [x] Request reviewers if possible

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/7153/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/7153/Hazelcast-pr-builder-7153.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast-archunit-rules: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 7 errors reported by Checkstyle 8.38 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-archunit-rules/src/main/java/com/hazelcast/test/archunit/ArchUnitRules.java:1: Line does not match expected header line of '/*'. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-archunit-rules/src/main/java/com/hazelcast/test/archunit/ArchUnitRules.java:1: Missing package-info.java file. [JavadocPackage]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-archunit-rules/src/main/java/com/hazelcast/test/archunit/ArchUnitRules.java:21:5: Missing a Javadoc comment. [JavadocVariable]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-archunit-rules/src/main/java/com/hazelcast/test/archunit/ArchUnitRules.java:47: Line is longer than 130 characters (found 146). [LineLength]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-archunit-rules/src/test/java/com/example/broken/BrokenSerializable.java:1: Missing a header - not enough lines in file. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-archunit-rules/src/test/java/com/example/valid/ValidSerializable.java:1: Missing a header - not enough lines in file. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-archunit-rules/src/test/java/com/hazelcast/test/archunit/ArchUnitRulesTest.java:1: Line does not match expected header line of '/*'. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/jet/impl/pipeline/transform/ProcessorTransform.java:46:5: Variable access definition in wrong order. [DeclarationOrder]
--------------------------
</pre>
</details>

frant-hartm commented on hazelcast/src/main/java/com/hazelcast/jet/impl/pipeline/transform/ProcessorTransform.java:
```diff
@@ -42,6 +42,7 @@
 
 public class ProcessorTransform extends AbstractTransform {
     public static final int NON_COOPERATIVE_DEFAULT_LOCAL_PARALLELISM = 2;
+    private static final long serialVersionUID = 1L;
```
It's a minor style thing but I would
- put the `serialVersionUID` field always first, so it's more visible if anyone is making changes
- separate it with empty line before & after

ldziedziul replied:
Yes, I wanted to put it first but unfortunately, checkstyle has a different opinion
```
[ERROR] /Users/ldziedziul/projects/hazelcast/hazelcast/hazelcast/src/main/java/com/hazelcast/jet/impl/pipeline/transform/ProcessorTransform.java:46:5: Variable access definition in wrong order. [DeclarationOrder]
```
Empty lines added here: https://github.com/hazelcast/hazelcast/pull/19534/commits/ddfbf33b32e0c0639dd8f05b925b00f4eed1e4e1

frant-hartm commented on hazelcast-archunit-rules/pom.xml:
```diff
@@ -0,0 +1,52 @@
+<!--
```
Will this module be deployed to maven central? If so it needs javadoc jar. If there is no class with javadoc then we already have a workaround for some jars without any javadoc which generates an empty jar. Search for `empty-javadoc-jar`

ldziedziul replied:
I think we don't have to deploy it, as it's only used during the build time. How do we control it?

frant-hartm replied:
Just realized it will eventually also be used in enterprise, so needs to be deployed.

ldziedziul replied:
added: https://github.com/ldziedziul/hazelcast/commit/8a7802761d56f1f07b2d2fd92dc36cec180b24a9

frant-hartm reviewed (changes requested):
We can either merge this after addressing the feedback or go even further and fix it for `com.hazelcast.jet` (add a test similar to SerializationTest in kinesis module) or other modules.

Up to you.

ldziedziul commented on hazelcast-archunit-rules/src/main/java/com/hazelcast/test/archunit/ArchUnitRules.java:
```diff
@@ -0,0 +1,75 @@
+            .and().doNotHaveModifier(JavaModifier.ABSTRACT)
+            .and().implement(Serializable.class)
+            .and().doNotImplement("com.hazelcast.nio.serialization.DataSerializable")
+            .and().areNotAssignableTo(Exception.class)
```
Good catch, a leftover after my experiments, fixed here: https://github.com/hazelcast/hazelcast/pull/19534/commits/f1984f01641716e2db68c50c435756b103647788

ldziedziul commented on extensions/kinesis/src/test/java/com/hazelcast/jet/kinesis/SerializableTest.java:
```diff
@@ -0,0 +1,40 @@
+import static com.tngtech.archunit.core.importer.ImportOption.Predefined.DO_NOT_INCLUDE_JARS;
+import static com.tngtech.archunit.core.importer.ImportOption.Predefined.DO_NOT_INCLUDE_TESTS;
+
+public class SerializableTest {
```
BTW should I mark the test with these annotations?
```
@RunWith(HazelcastParallelClassRunner.class)
@Category({QuickTest.class, ParallelJVMTest.class})
```

What's the heuristic of applying them?

frant-hartm replied:
Use `HazelcastParallelClassRunner` for running methods in parallel inside single jvm - such tests must not use real network with conflicting ports and must not use any static resources, basically perfectly isolated

use `@Category(ParallelJVMTest.class)` to run test classes in parallel with other test classes (in different jvm) - these must not use real network, but can use static resources because they are inside a separate jvm

`QuickTest` was used to mark test for the PR builder, now when there is no annotation it is automatically. There is a pending PR to remove it, but there was some issue with test hierarchies I think.

ldziedziul commented:
> We can either merge this after addressing the feedback or go even further

I would like to merge it since it seems that the jet module can be more tricky.

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/7199/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/7199/Hazelcast-pr-builder-7199.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   TestEmptyApplicationContext.testXmlWithoutConfig ? IllegalState Failed to load...
[INFO] 
[ERROR] Tests run: 201, Failures: 0, Errors: 1, Skipped: 0
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

ldziedziul commented:
> I would like to merge it since it seems that the jet module can be more tricky.

Well, I've changed my mind, it wasn't that complicated (probably ;) )

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/7211/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/7211/Hazelcast-pr-builder-7211.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast-jet-cdc-mysql: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.38 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/extensions/cdc-mysql/../../checkstyle/checkstyle_jet.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/extensions/cdc-mysql/src/test/java/com/hazelcast/jet/cdc/mysql/CdcMysqlSerializableTest.java:2: Line does not match expected header line of ' * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.'. [Header]
--------------------------
</pre>
</details>

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/7212/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/7212/Hazelcast-pr-builder-7212.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   CheckDependenciesIT.testNoMandatoryDependencyDeclared:77->getHazelcastManifest:105->findHazelcastManifestURL:147 Found multiple matching URLs: jar:file:/home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-archunit-rules/target/hazelcast-archunit-rules-5.0-SNAPSHOT.jar!/META-INF/MANIFEST.MF and jar:file:/home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/target/hazelcast-5.0-SNAPSHOT.jar!/META-INF/MANIFEST.MF
[ERROR]   CheckDependenciesIT.verifyManifestEntries:94->getHazelcastManifest:105->findHazelcastManifestURL:147 Found multiple matching URLs: jar:file:/home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast-archunit-rules/target/hazelcast-archunit-rules-5.0-SNAPSHOT.jar!/META-INF/MANIFEST.MF and jar:file:/home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/target/hazelcast-5.0-SNAPSHOT.jar!/META-INF/MANIFEST.MF
[INFO] 
[ERROR] Tests run: 5, Failures: 2, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- maven-failsafe-plugin:2.22.2:verify (default) @ hazelcast ---
[INFO] Failsafe report directory: /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/target/failsafe-reports

[ERROR] There are test failures.
</pre>
</details>

frant-hartm commented on extensions/elasticsearch/elasticsearch-5/src/main/java/com/hazelcast/jet/elastic/ElasticSinkBuilder.java:
```diff
@@ -62,6 +62,8 @@
  */
 public final class ElasticSinkBuilder<T> implements Serializable {
 
+    private static final long serialVersionUID = 1L;
```
This class actually shouldn't be `Serializable` (also version 6. and 7). 

I will send a separate PR with a fix.

ldziedziul commented on pom.xml:
```diff
@@ -106,6 +107,7 @@
         <kafka.version>2.2.2</kafka.version>
         <kotlin.version>1.3.61</kotlin.version>
         <log4j.version>1.2.17</log4j.version>
+        <log4j2.slf4j.binding.version>2.13.2</log4j2.slf4j.binding.version>
```
Sure, I'll unify them in another PR

frant-hartm replied:
I meant to take these changes out of this PR (and send it in another one).

ldziedziul commented on hazelcast/src/main/java/com/hazelcast/jet/core/EventTimePolicy.java:
```diff
@@ -70,11 +70,14 @@
     private static final SupplierEx<WatermarkPolicy> NO_WATERMARKS = () -> new WatermarkPolicy() {
         @Override
-        public void reportEvent(long timestamp) { }
+        public void reportEvent(long timestamp) {
```
In general, I agree with you, My IDE reformats parts of the files I've changed. It seems that my changes were too close to this constructor. Treat this please as an exceptional reformatting 😇

ldziedziul commented on hazelcast/src/main/java/com/hazelcast/jet/impl/processor/HashJoinCollectP.java:
```diff
@@ -88,6 +88,9 @@ public boolean complete() {
     // ArrayList and then the logic that relies on instanceof would break
     static final class HashJoinArrayList extends ArrayList<Object> {
+
+        private static final long serialVersionUID = 1L;
```
Yes, we can, although does it hurt to have here?

frant-hartm replied:
Not really.

mdumandag commented on hazelcast/src/test/java/com/hazelcast/jet/JetSerializableTest.java:
```diff
@@ -0,0 +1,38 @@
+import static com.tngtech.archunit.core.importer.ImportOption.Predefined.DO_NOT_INCLUDE_JARS;
+import static com.tngtech.archunit.core.importer.ImportOption.Predefined.DO_NOT_INCLUDE_TESTS;
+
+public class JetSerializableTest {
```
Do we need annotations here?

ldziedziul replied:
yes! thanks, fixed: https://github.com/hazelcast/hazelcast/pull/19534/commits/c341fdf311ae773729cc715b9e8ae7aa22b71132

frant-hartm replied:
Good catch. This should have the ParallelJVMTest category.

mdumandag reviewed (commented):
Looks OK from the client's perspective. Just left a one comment

frant-hartm reviewed (approved):
Two minor things left.

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/7444/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/7444/Hazelcast-pr-builder-7444.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   ExpirationTimeTest.expire_based_on_ttl_when_when_ttl_is_smaller_than_max_idle:94->getExpirationTime:668 NullPointer
[INFO] 
[ERROR] Tests run: 42942, Failures: 0, Errors: 1, Skipped: 1011
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

ldziedziul commented:
run-lab-run

vbekiaris reviewed (commented):
post-merge 👍  , thanks @ldziedziul for the PR
